`nats_consumer_num_pending` metric reporting different numbers across nodes

### Observed behavior

Timeline:
1. Consumer processes all messages (68k+ total messages in stream). All looking healthy. `nats_consumer_num_pending` reports 0 correctly.
2. Consumer config is updated: BackOff=[5s, 15s, 30s], MaxDeliver bumped from 3 → 4
3. Immediately after the update, `nats_consumer_num_pending` metric jumps to 67k+.
4. The value in the metrics stays flat for a couple of hours until the consumer was manually recreated with a new leader. However, the leader (via `nats consumer info`) shows 0 pending throughout the duration.

The graph reports incorrect number ~11am-1.30pm. However label`consumer_leader="nats-1"` is correctly reported as `nats-1` was the leader at the time.
<img width="2155" height="358" alt="Image" src="https://github.com/user-attachments/assets/b33e18c3-9c0a-4af0-bf9f-81385f9217ae" />

During my investigation, I've found [this issue](https://github.com/nats-io/nats-server/issues/7838) reported previously that seems to have similar behaviours. It seems that the follower nodes take a while to sync, and if the prometheus scrape request is sent to a follower node, then the metrics can report stale numbers.

Would like to confirm and check if this is an expected behavior or is concerning.



### Expected behavior

`nats_consumer_num_pending` would stay at 0 after the consumer config update. Or a shorter lag (instead of a couple of hours) for replicas to sync up and report the correct number.

### Server and client version

NATS Server: v2.12.1
Cluster: 3-node JetStream
Consumer: durable pull, DeliverAll, AckExplicit
Stream: large (67k+ messages at time of config update)

Stream info:
```
~ # nats stream info test_stream
Information for Stream test_stream created 2026-02-13 16:45:21

                Subjects: test.something
                Replicas: 3
                 Storage: File

Options:

               Retention: Limits
         Acknowledgments: true
          Discard Policy: Old
        Duplicate Window: 2m0s
    Allows Batch Publish: false
         Allows Counters: false
       Allows Msg Delete: true
  Allows Per-Message TTL: false
            Allows Purge: true
        Allows Schedules: false
          Allows Rollups: false

Limits:

        Maximum Messages: unlimited
     Maximum Per Subject: unlimited
           Maximum Bytes: unlimited
             Maximum Age: 7d0h0m0s
    Maximum Message Size: unlimited
       Maximum Consumers: unlimited

Cluster Information:

                    Name: nats
           Cluster Group: S-R1F-1s775bGH
                  Leader: nats-1 (6d0h24m51s)
                 Replica: nats-0, current, seen 238ms ago
                 Replica: nats-2, current, seen 238ms ago

State:

            Host Version: 2.12.1
      Required API Level: 0 hosted at level 2
                Messages: 68,824
                   Bytes: 62 MiB
          First Sequence: 1 @ 2026-05-08 17:43:05
           Last Sequence: 68,824 @ 2026-05-11 17:35:22
        Active Consumers: 1
      Number of Subjects: 1
```

Consumer info at the time of incident:
```
~ # nats consumer info test_stream test_consumer
Information for Consumer test_stream > test_consumer created 2026-02-13 16:45:14

Configuration:

                    Name: test_consumer
               Pull Mode: true
         Filter Subjects: test.something
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 5.00s
           Replay Policy: Instant
      Maximum Deliveries: 4
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512
                 Backoff: 5s, 15s, 30s

Cluster Information:

                    Name: nats
              Raft Group: C-R1F-eusUVnjl
                  Leader: nats-1 (6d0h18m46s)
                 Replica: nats-0, current, seen 64ms ago
                 Replica: nats-2, current, seen 64ms ago

State:

            Host Version: 2.12.1
      Required API Level: 0 hosted at level 2
  Last Delivered Message: Consumer sequence: 69,039 Stream sequence: 68,984 Last delivery: 7.31s ago
    Acknowledgment Floor: Consumer sequence: 69,039 Stream sequence: 68,984 Last Ack: 7.06s ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 26
    Unprocessed Messages: 0
           Waiting Pulls: 6 of maximum 512
```


### Host environment

_No response_

### Steps to reproduce

_No response_

> Could you share how to reproduce this?

Here's the code snippet that I used to test the behavior (drop the file into the `server/` directory of this repo): 

https://gist.github.com/tian-pc/d1ccdfc936b4cf0d2719c52b52d90a4a


The test fails with output:

```
=== RUN   TestJetStreamClusterConsumerNumPendingFollowerBug
    jetstream_consumer_backoff_num_pending_test.go:119: Before update — leader S-3: NumPending=0, follower S-1: NumPending=0
    jetstream_consumer_backoff_num_pending_test.go:138: After update  — leader S-3: NumPending=0
    jetstream_consumer_backoff_num_pending_test.go:139: After update  — follower S-1: NumPending=500 (want 0, got 500 if bug present)
    jetstream_consumer_backoff_num_pending_test.go:144: Scrape 1 — follower S-1: NumPending=500
    jetstream_consumer_backoff_num_pending_test.go:144: Scrape 2 — follower S-1: NumPending=500
    jetstream_consumer_backoff_num_pending_test.go:144: Scrape 3 — follower S-1: NumPending=500
    jetstream_consumer_backoff_num_pending_test.go:144: Scrape 4 — follower S-1: NumPending=500
    jetstream_consumer_backoff_num_pending_test.go:144: Scrape 5 — follower S-1: NumPending=500
    jetstream_consumer_backoff_num_pending_test.go:147: require uint64 equal, but got: 500 != 0
--- FAIL: TestJetStreamClusterConsumerNumPendingFollowerBug (1.61s)
```

Follower `S-1` has 0 pending before the consumer update, but jumped to 500 after.

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -1838,6 +1838,8 @@ func (o *consumer) setLeader(isLeader bool) error {
 		o.pending = nil
 		o.rsm = nil
 		o.resetPendingDeliveries()
+		// Reset num pending, these are only authoritative on the leader.
+		o.npc, o.npf = 0, 0
 		// ok if they are nil, we protect inside unsubscribe()
 		o.unsubscribe(o.ackSubOld)
 		o.unsubscribe(o.ackSub)

--- a/server/consumer.go
+++ b/server/consumer.go
@@ -5544,11 +5544,11 @@ func (o *consumer) streamNumPendingLocked() (uint64, error) {
 	return o.streamNumPending()
 }
 
-// Will force a set from the stream store of num pending.
+// Will force a set from the stream store of num pending on the consumer leader.
 // Depends on delivery policy, for last per subject we calculate differently.
 // Lock should be held.
 func (o *consumer) streamNumPending() (uint64, error) {
-	if o.mset == nil || o.mset.store == nil {
+	if o.mset == nil || o.mset.store == nil || !o.isLeader() {
 		o.npc, o.npf = 0, 0
 		return 0, nil
 	}

--- a/server/jetstream_cluster.go
+++ b/server/jetstream_cluster.go
@@ -10265,9 +10265,7 @@ func (mset *stream) processSnapshot(snap *StreamReplicatedState, index uint64) (
 		mset.mu.Lock()
 		for _, o := range mset.consumers {
 			o.mu.Lock()
-			if o.isLeader() {
-				o.streamNumPending()
-			}
+			o.streamNumPending()
 			o.mu.Unlock()
 		}
 		mset.mu.Unlock()

--- a/server/jetstream_consumer_test.go
+++ b/server/jetstream_consumer_test.go
@@ -12288,3 +12288,56 @@ func TestJetStreamConsumerStreamNumPendingOnlyOnLeader(t *testing.T) {
 	require_Equal(t, ci.NumPending, 1)
 	checkNumPending(1)
 }
+
+// https://github.com/nats-io/nats-server/issues/8140
+func TestJetStreamConsumerStreamNumPendingClearedOnStepDown(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	nc, js := jsClientConnect(t, c.randomServer())
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"foo"},
+		Replicas: 3,
+	})
+	require_NoError(t, err)
+
+	for range 5 {
+		_, err = js.Publish("foo", nil)
+		require_NoError(t, err)
+	}
+
+	_, err = js.AddConsumer("TEST", &nats.ConsumerConfig{Durable: "CONSUMER"})
+	require_NoError(t, err)
+
+	oldLeader := c.consumerLeader(globalAccountName, "TEST", "CONSUMER")
+	require_NotNil(t, oldLeader)
+	mset, err := oldLeader.globalAccount().lookupStream("TEST")
+	require_NoError(t, err)
+	o := mset.lookupConsumer("CONSUMER")
+	require_NotNil(t, o)
+
+	// Sanity: leader has the computed value.
+	require_Equal(t, o.info().NumPending, 5)
+
+	// Step down the consumer leader.
+	n := o.raftNode()
+	require_NotNil(t, n)
+	require_NoError(t, n.StepDown())
+
+	// The old leader must transition to follower, and its in-memory NumPending must be reset.
+	checkFor(t, 5*time.Second, 100*time.Millisecond, func() error {
+		o.mu.RLock()
+		npc, isLeader := o.npc, o.isLeader()
+		o.mu.RUnlock()
+		if isLeader {
+			return fmt.Errorf("old leader %s has not stepped down", oldLeader.Name())
+		}
+		if npc != 0 {
+			return fmt.Errorf("old leader %s still reports stale npc=%d, expected 0", oldLeader.Name(), npc)
+		}
+		return nil
+	})
+}

--- a/server/jetstream_consumer_test.go
+++ b/server/jetstream_consumer_test.go
@@ -12219,3 +12219,72 @@ func TestJetStreamConsumerStepDownResetsPendingAndRdc(t *testing.T) {
 		return nil
 	})
 }
+
+// https://github.com/nats-io/nats-server/issues/8140
+func TestJetStreamConsumerStreamNumPendingOnlyOnLeader(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	nc, js := jsClientConnect(t, c.randomServer())
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"foo.*"},
+		Replicas: 3,
+	})
+	require_NoError(t, err)
+
+	for _, subj := range []string{"a", "b"} {
+		_, err = js.Publish(fmt.Sprintf("foo.%s", subj), nil)
+		require_NoError(t, err)
+	}
+
+	cfg := &nats.ConsumerConfig{Durable: "CONSUMER"}
+	ci, err := js.AddConsumer("TEST", cfg)
+	require_NoError(t, err)
+	require_Equal(t, ci.NumPending, 2)
+
+	checkNumPending := func(expected uint64) {
+		t.Helper()
+		checkFor(t, 2*time.Second, 200*time.Millisecond, func() error {
+			for _, s := range c.servers {
+				mset, err := s.globalAccount().lookupStream("TEST")
+				if err != nil {
+					return err
+				}
+				o := mset.lookupConsumer("CONSUMER")
+				if o == nil {
+					return fmt.Errorf("consumer not found on %s", s.Name())
+				}
+				ici := o.info()
+				if o.IsLeader() {
+					if ici.NumPending != expected {
+						return fmt.Errorf("leader %s: expected NumPending=%d, got %d", s.Name(), expected, ici.NumPending)
+					}
+				} else if ici.NumPending != 0 {
+					return fmt.Errorf("follower %s: expected NumPending=0, got %d", s.Name(), ici.NumPending)
+				}
+			}
+			return nil
+		})
+	}
+	checkNumPending(2)
+
+	ci, err = js.ConsumerInfo("TEST", "CONSUMER")
+	require_NoError(t, err)
+	require_Equal(t, ci.NumPending, 2)
+	checkNumPending(2)
+
+	// Update the filters, which should recalculate num pending, but only on the leader.
+	cfg.FilterSubject = "foo.b"
+	ci, err = js.UpdateConsumer("TEST", cfg)
+	require_NoError(t, err)
+	require_Equal(t, ci.NumPending, 1)
+	checkNumPending(1)
+
+	ci, err = js.ConsumerInfo("TEST", "CONSUMER")
+	require_NoError(t, err)
+	require_Equal(t, ci.NumPending, 1)
+	checkNumPending(1)
+}

